# Example Gradient Boosting

In this example, we will work with the wine classification dataset, which is built into Scikit-Learn

In [1]:
from sklearn.datasets import load_wine

In [2]:
wine = load_wine() # Initializing the wine data
wine # Full description of the dataset

{'data': array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
         1.065e+03],
        [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
         1.050e+03],
        [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
         1.185e+03],
        ...,
        [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
         8.350e+02],
        [1.317e+01, 2.590e+00, 2.370e+00, ..., 6.000e-01, 1.620e+00,
         8.400e+02],
        [1.413e+01, 4.100e+00, 2.740e+00, ..., 6.100e-01, 1.600e+00,
         5.600e+02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1

In [3]:
print(wine.feature_names) # Printing the features
print(wine.target_names) # Printing the labels

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']
['class_0' 'class_1' 'class_2']


#


# Separating the data into features and target

In [4]:
X = wine.data # Feature
y = wine.target # Target

# Printing the first 5 rows
print(X[0:5])
print('-------')
print(y[0:5])

[[1.423e+01 1.710e+00 2.430e+00 1.560e+01 1.270e+02 2.800e+00 3.060e+00
  2.800e-01 2.290e+00 5.640e+00 1.040e+00 3.920e+00 1.065e+03]
 [1.320e+01 1.780e+00 2.140e+00 1.120e+01 1.000e+02 2.650e+00 2.760e+00
  2.600e-01 1.280e+00 4.380e+00 1.050e+00 3.400e+00 1.050e+03]
 [1.316e+01 2.360e+00 2.670e+00 1.860e+01 1.010e+02 2.800e+00 3.240e+00
  3.000e-01 2.810e+00 5.680e+00 1.030e+00 3.170e+00 1.185e+03]
 [1.437e+01 1.950e+00 2.500e+00 1.680e+01 1.130e+02 3.850e+00 3.490e+00
  2.400e-01 2.180e+00 7.800e+00 8.600e-01 3.450e+00 1.480e+03]
 [1.324e+01 2.590e+00 2.870e+00 2.100e+01 1.180e+02 2.800e+00 2.690e+00
  3.900e-01 1.820e+00 4.320e+00 1.040e+00 2.930e+00 7.350e+02]]
-------
[0 0 0 0 0]



# Separating the data into training and test

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.2, # Separating 20% of the data for the test set
                                                    random_state=42 # For replicating results
                                                    )

# Importing Gradient Boosting

In [6]:
from sklearn.ensemble import GradientBoostingClassifier # Because we are classifying wines into 0, 1, and 2

gb = GradientBoostingClassifier(n_estimators=100, # number of decision trees
                                learning_rate=0.1, # Learning rate
                                max_depth=1, # Tree depth
                                random_state=42
                                ).fit(X_train, y_train)



# Making predictions

In [7]:
y_pred = gb.predict(X_test)

# Evaluating the score

In [8]:
gb.score(X_test, y_test)

1.0

### Although this may sound wonderful, we have to be skeptical and consider whether the model simply memorized the data, leading to overfitting

### Let's analyze the Confusion Matrix and the classification report

In [9]:
from sklearn.metrics import confusion_matrix, classification_report

In [10]:
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[14  0  0]
 [ 0 14  0]
 [ 0  0  8]]


In [11]:
print(classification_report(y_test, y_pred, target_names=wine.target_names))

              precision    recall  f1-score   support

     class_0       1.00      1.00      1.00        14
     class_1       1.00      1.00      1.00        14
     class_2       1.00      1.00      1.00         8

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



Indeed, the model got every prediction right, correctly classifying all the test data

# Performing Cross-Validation to confirm overfitting and achieve a more reliable model performance

In [12]:
from sklearn.model_selection import cross_val_score

We divided the dataset into 3 chunks, training and testing the model multiple times, using different chunks for training and testing for each iteration, retrieving the mean score for each fold

In [13]:

cross_val_score(gb, X_train, y_train, cv=3, n_jobs=-1).mean()

np.float64(0.9645390070921986)

# Searching for the best hyperparameters

In [14]:
from sklearn.model_selection import GridSearchCV

In [15]:
param_grid = {
    'n_estimators': [10, 50, 100, 500],
    'learning_rate': [0.0001, 0.001, 0.01, 1.0],
    'max_depth': [3, 5, 7, 9]
}

# Combining each parameter from the object

In [16]:
gb2 = GridSearchCV(gb, param_grid, cv=3, n_jobs=-1)

In [17]:
gb2.fit(X_train, y_train)

GridSearchCV(cv=3,
             estimator=GradientBoostingClassifier(max_depth=1, random_state=42),
             n_jobs=-1,
             param_grid={'learning_rate': [0.0001, 0.001, 0.01, 1.0],
                         'max_depth': [3, 5, 7, 9],
                         'n_estimators': [10, 50, 100, 500]})

In [18]:
gb2.best_params_ # The best combination of hyperparameters

{'learning_rate': 1.0, 'max_depth': 3, 'n_estimators': 10}

In [19]:
gb2.best_score_ # Best score

np.float64(0.9716312056737588)

In [20]:
best_gb_model = gb2.best_estimator_

y_pred_best = best_gb_model.predict(X_test)

cm_best = confusion_matrix(y_test, y_pred_best)

In [21]:
print(cm_best)

[[14  0  0]
 [ 0 14  0]
 [ 0  1  7]]


# Conclusion
The inherent datasets in Scikit-Learn are well-behaved; that is, any model will provide an excellent result, as our 1.0 score indicates. The mean score with cross-validation for the 3-chunk division (0.96) shows that the model didn't necessarily memorize all the data, since it used different subsets for testing and training.

In the search for the best hyperparameters, the best score (0.97) reinforces the thesis that the model generalized well for new data, without overfitting.